File: src/main/js/formulas/math/ASIN.ts
import { NumException } from "../../errors/NumException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class ASIN extends AbstractFormula {
  static SELF: ASIN = new ASIN();
  NAME = FormulaName.ASIN;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const value = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[0])));
    if (value < -1 || value > 1) {
      return new NumException(
        `ASIN parameter 1 value is ${value}. Valid values should be between -1 and 1, inclusively`
      );
    }
    return Math.asin(value);
  }
}

File: src/main/java/io/protobase/f7/formulas/math/ASIN.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class ASIN extends AbstractFormula {
  public static FormulaName NAME = FormulaName.ASIN;
  public static ASIN SELF = new ASIN();

  public ASIN() {
    super();
  }

  public ASIN(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    return Math.asin(Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[0]))));
  }
}
